core/resharding.go

```diff
@@ -0,0 +1,173 @@
+package core
+
+import (
+	"math/rand"
+	"sort"
+	"strconv"
+
+	"github.com/harmony-one/harmony/core/types"
+)
+
+const (
+	// InitialSeed is the initial random seed, a magic number to answer everything, remove later
+	InitialSeed int64 = 42
+)
+
+// ShardingState is data structure hold the sharding state
+type ShardingState struct {
+	epoch      uint64 // current epoch
+	rnd        int64  // random seed for resharding
+	numShards  int
+	shardState types.ShardState
+}
+
+// sortedCommitteeBySize will sort shards by size
+// Suppose there are N shards, the first N/2 larger shards are called active committees
+// the rest N/2 smaller committees are called inactive committees
+// actually they are all just normal shards
+func (ss *ShardingState) sortCommitteeBySize() {
+	sort.Slice(ss.shardState, func(i, j int) bool {
+		return len(ss.shardState[i].NodeList) > len(ss.shardState[j].NodeList)
+	})
+}
+
+// assignNewNodes add new nodes into the N/2 active committees evenly
+func (ss *ShardingState) assignNewNodes(newNodeList []types.NodeID) {
+	ss.sortCommitteeBySize()
+	numActiveShards := ss.numShards / 2
+	Shuffle(newNodeList)
+	for i, nid := range newNodeList {
+		id := i % numActiveShards
+		ss.shardState[id].NodeList = append(ss.shardState[id].NodeList, nid)
+	}
+}
+
+// cuckooResharding uses cuckoo rule to reshard X% of active committee(shards) into inactive committee(shards)
+func (ss *ShardingState) cuckooResharding(percent float64) {
+	ss.sortCommitteeBySize()
+	numActiveShards := ss.numShards / 2
+	kickedNodes := []types.NodeID{}
+	for i := range ss.shardState {
+		if i >= numActiveShards {
+			break
+		}
+		Shuffle(ss.shardState[i].NodeList)
+		numKicked := int(percent * float64(len(ss.shardState[i].NodeList)))
+		tmp := ss.shardState[i].NodeList[:numKicked]
+		kickedNodes = append(kickedNodes, tmp...)
+		ss.shardState[i].NodeList = ss.shardState[i].NodeList[numKicked:]
+	}
+
+	Shuffle(kickedNodes)
+	for i, nid := range kickedNodes {
+		id := numActiveShards + i%(ss.numShards-numActiveShards)
+		ss.shardState[id].NodeList = append(ss.shardState[id].NodeList, nid)
+	}
+}
+
+// UpdateShardState will first add new nodes into shards, then use cuckoo rule to reshard to get new shard state
+func (ss *ShardingState) UpdateShardState(newNodeList []types.NodeID, percent float64) {
+	rand.Seed(ss.rnd)
+	ss.cuckooResharding(percent)
+	ss.assignNewNodes(newNodeList)
```

Comment:
I think the algorithm start with assigning new node to active shards, and then do cuckoo resharding (with newly added nodes). So new nodes may up end in the inactive shards.

Reply:
I thought we want to the new nodes stay in active shards. In this case, I will change the order of above lines